domain/resource/state/resource.go

```diff
@@ -1356,6 +1356,191 @@ WHERE uuid = $updatePotentialResource.uuid
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	// Does the resource name exist for the charm... get UUID of
+	// Prepare SQL statement to insert the resource.
+	insertStmt, err := st.Prepare(`
+INSERT INTO resource (uuid, charm_uuid, charm_resource_name, revision, 
+       origin_type_id, state_id, created_at)
+SELECT $resourceToAdd.uuid,
+       $resourceToAdd.charm_uuid,
+       $resourceToAdd.charm_resource_name,
+       $resourceToAdd.revision,
+       rot.id,
+       rs.id,
+       $resourceToAdd.created_at
+FROM   resource_origin_type rot,
+       resource_state rs
+WHERE  rot.name = $resourceToAdd.origin_type_name
+AND    rs.name = $resourceToAdd.state_name`, resourceToAdd{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	// Prepare SQL statement to link resource with application name.
+	linkStmt, err := st.Prepare(`
+INSERT INTO pending_application_resource (application_name, resource_uuid)
+VALUES ($linkResourceApplication.*)`, linkResourceApplication{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Insert resources
+		for _, res := range resources {
+			// Insert the resource.
+			if err = tx.Query(ctx, insertStmt, res).Run(); err != nil {
+				return errors.Errorf("inserting resource %q: %w", res.Name, err)
+			}
+			// Link the resource to the application name.
+			if err = tx.Query(ctx, linkStmt, linkResourceApplication{
+				ResourceUUID:    res.UUID,
+				ApplicationName: args.ApplicationName,
+			}).Run(); err != nil {
+				return errors.Errorf(
+					"linking resource %q to application %q: %w",
+					res.Name, args.ApplicationName, err)
+			}
+		}
+		return nil
+	})
+	return resourceUUIDs, errors.Capture(err)
+}
+
+// buildResourceInserts creates resources to add based on provided app and charm
+// resources.
+// Returns a slice of resourcesToAdd, a slice of the created resource UUIDs, and
+// an error if any issues occur during creation.
+func (st *State) buildResourcesToAdd(
+	charmUUID string,
+	appResources []resource.AddResourceDetails,
+) ([]resourceToAdd, []coreresource.UUID, error) {
+	resources := make([]resourceToAdd, len(appResources))
+	result := make([]coreresource.UUID, len(appResources))
+	now := st.clock.Now()
+	for i, r := range appResources {
+		uuid, err := coreresource.NewUUID()
+		if err != nil {
+			return nil, nil, errors.Capture(err)
+		}
+		result[i] = uuid
+		resources[i] = resourceToAdd{
+			UUID:      uuid.String(),
+			CharmUUID: charmUUID,
+			Name:      r.Name,
+			Revision:  r.Revision,
+			Origin:    r.Origin.String(),
+			State:     coreresource.StateAvailable.String(),
+			CreatedAt: now,
+		}
+	}
+	return resources, result, nil
+}
+
+// UpdateStoreResource updates the revision of a store resource to a new
+// version. Increments charm modified version for the application to
+// trigger use of the new resource revision by the application.
+func (st *State) UpdateStoreResource(
```

Comment:
See comment on the service method, I think you can get away with not adding this method.